### PR TITLE
Refactor logging to provide common logger from LoggerAdapter

### PR DESCRIPTION
**Pull request by flovilmart**
- Refactors logger into the WinstonLoggerAdapter and WinstonLogger
- Exposes the loggerController, instead of the adapter
- adds `silent` option to silent all console logs (from command line or code)
- adds `logLevel / PARSE_SERVER_LOG_LEVEL` option for setting the log level
- Custom logger adapters will be passed `logLevel, logsFolder, silent, verbose, jsonLogs` options
- Custom loggers don't require the query function anymore to allow fan out loggers
- Custom loggers require only `log` to load properly 
- LoggerController exposes more methods (cf: [winston](https://github.com/winstonjs/winston#logging-levels) and [npm](https://github.com/npm/npmlog/blob/master/log.js)))

### CODE REVIEW

File: src/Adapters/Logger/LoggerAdapter.js [hunk 1]
```diff
@@ -6,12 +6,15 @@
...
 // * query(options, callback)
+// * get logger()
+// * configureLogger(options)
 // Default is WinstonLoggerAdapter.js
 
 export class LoggerAdapter {
   info() {}
   error() {}
   query(options, callback) {}
+  configureLogger(options) {}
```
Comment (steven-supersolid) [inline src/Adapters/Logger/LoggerAdapter.js]: Should we include an empty get logger() here/

File: src/logging.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    console.log('NO LOGGER SET!');
+  }
+  logger = aLogger;
+}
+
+function getLogger() {
+  return logger;
+}
+
+export function addTransport(transport) {
```
Comment (steven-supersolid) [inline src/logging.js]: Is this still useful? I think if we set a different logger then it won't do what's expected. Could we instead move inside a logger adapter implementation? Also, transport seems like a winston specific feature

File: src/logging.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  getLogger,
+  addTransport
+};
+
+Object.defineProperty(logging, 'logger', {
+  get: getLogger
+});
+
+module.exports = logging;
+export default logging;
```
Comment (steven-supersolid) [inline src/logging.js]: I don't know if this is possible but may be more convenient to export logger/getLogger as the default as that is used most often and export logging (or just setLogger) as something else.

File: src/logging.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+function getLogger() {
+  return logger;
+}
+
+export function addTransport(transport) {
+  winstonLogger.addTransport(transport);
+}
+
+let logging = {
+  setLogger,
```
Comment (flovilmart) [inline src/logging.js]: What do you think @acinader

File: src/logging.js (hunk 2 above)
Comment (flovilmart) [inline src/logging.js]: What do you think @acinader

File: src/logging.js (hunk 3 above)
Comment (flovilmart) [inline src/logging.js]: Yeah, that may remove some code

File: src/Routers/FunctionsRouter.js [hunk 5]
```diff
@@ -78,14 +78,14 @@ export class FunctionsRouter extends PromiseRouter {
 
       return new Promise(function (resolve, reject) {
         var response = FunctionsRouter.createResponseObject((result) => {
-          logger.info(`Ran cloud function ${req.params.functionName} with:\nInput: ${JSON.stringify(params)}\nResult: ${JSON.stringify(result.response.result)}`, {
+          log.logger.info(`Ran cloud function ${req.params.functionName} with:\nInput: ${JSON.stringify(params)}\nResult: ${JSON.stringify(result.response.result)}`, {
```
Comment (acinader) [inline src/Routers/FunctionsRouter.js]: log.logger.info seems unfortunate.  any way we can do without that level indirection?

File: src/logging.js (hunk 2 above)
Comment (acinader) [inline src/logging.js]: it sure doesn't seem to belong there anymore.  src/Adapters/Logger/WinstonLogger.js handles the transport collection after this pr, so it would seem that the ability to add a transport would go in there if it is to be exposed at all.

there is a likely an easier way to do what i did, but the goal of my change was to allow me to add a transport to the existing logger.

my first thought would be to move additional transports into configureLogger, the transport could be passed as an option.

Comment (flovilmart) [inline src/logging.js]: I can move forward additional logger options to the initializer. Or give an optional callback with the full winstonAdapter

Comment (acinader) [inline src/logging.js]: just to make sure i am following, it isn't the winstonAdapter that handles the transports, its WinstonLogger, right?

Comment (acinader) [inline src/logging.js]: and there's a getLogger(), so I could log.getLogger().addTransport(), right?

Comment (flovilmart) [inline src/logging.js]: Yes that's the Winston logger.

But something's off... i'll try to do better as it's awefully confusing.

File: src/ParseServer.js [hunk 6]
```diff
@@ -177,16 +174,22 @@ class ParseServer {
...
-
     const filesControllerAdapter = loadAdapter(filesAdapter, () => {
       return new GridStoreAdapter(databaseURI);
     });
     // Pass the push options too as it works with the default
     const pushControllerAdapter = loadAdapter(push && push.adapter, ParsePushAdapter, push || {});
+    
     const loggerControllerAdapter = loadAdapter(loggerAdapter, WinstonLoggerAdapter);
+
+    log.setLogger(loggerControllerAdapter.logger);
```
Comment (steven-supersolid) [inline src/ParseServer.js]: Because we set logger here, could we move earlier and avoid the default logger in logging.js? Or are we always going to need a default logger?

File: src/Adapters/Logger/WinstonLoggerAdapter.js [hunk 7]
```diff
@@ -60,6 +60,14 @@ export class WinstonLoggerAdapter extends LoggerAdapter {
...
+    return logger;
+  }
+
+  configureLogger(options) {
+    configureLogger(options);
+  }
+
   addTransport(transport) {
     // Note that this is calling addTransport
     // from logger.  See import - confusing.
```
Comment (steven-supersolid) [inline src/Adapters/Logger/WinstonLoggerAdapter.js]: Can we make this not confusing, e.g. call logger.addTransport() or similar instead?

File: src/Adapters/Logger/WinstonLoggerAdapter.js [hunk 8]
```diff
@@ -1,5 +1,5 @@
 import { LoggerAdapter } from './LoggerAdapter';
-import { logger, addTransport } from '../../logger';
+import { logger, addTransport, configureLogger } from './WinstonLogger';
```
Comment (steven-supersolid) [inline src/Adapters/Logger/WinstonLoggerAdapter.js]: Could we just move WinstonLogger functionality inside this class? May simplify things.
The main problem I see if we did that is creating the default logger in logging.js - one solution would be to create a WinstonLoggerAdapter in logging.js and call get logger() on that, instead of creating a WinstonLogger instance.

File: src/Adapters/Logger/WinstonLoggerAdapter.js [hunk 9]
```diff
@@ -60,6 +60,22 @@ export class WinstonLoggerAdapter extends LoggerAdapter {
...
 
+  warn() {
+    return logger.warn.apply(undefined, arguments);
+  }
+
+  verbose() {
+    return logger.verbose.apply(undefined, arguments);
+  }
+
+  log() {
```
Comment (acinader) [inline src/Adapters/Logger/WinstonLoggerAdapter.js]: https://github.com/ParsePlatform/parse-server/blob/fa736f1df7b6a4f6dbc3f08b2f9db8711de23694/src/ParseServer.js#L181

configured for silly?

I'd suggest we just implement the full sweet?  debug is useful.  https://github.com/winstonjs/winston#logging-levels

Comment (flovilmart) [inline src/Adapters/Logger/WinstonLoggerAdapter.js]: agreed. 

In winston, the 1st argument of log is the level you want to log to. 

As for silly, I can down to verbose as we don't log higher

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

File: spec/Logger.spec.js [hunk 10]
```diff
@@ -8,11 +8,15 @@ class TestTransport extends winston.Transport {
 }
 
 describe('Logger', () => {
+  // Test is excluded as will be refactored
```
Comment (acinader) [inline spec/Logger.spec.js]: remove

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-8** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-7** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-4** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-3** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@acinader, @steven-supersolid, @TylerBrock/@spenthil I'm good with the changes on that PR. I you guys want to review, feel free.

File: src/Adapters/Logger/WinstonLogger.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  winston.level = logLevel;
+  const options = {};
+
+  if (logsFolder) {
+    if (!path.isAbsolute(logsFolder)) {
+      logsFolder = path.resolve(process.cwd(), logsFolder);
+    }
+    try {
+      fs.mkdirSync(logsFolder);
+    } catch (exception) {}
```
Comment (acinader) [inline src/Adapters/Logger/WinstonLogger.js]: output to console or don't catch?

File: src/ParseServer.js [hunk 12]
```diff
@@ -177,16 +174,13 @@ class ParseServer {
...
-      configureLogger({level: 'silly', jsonLogs});
-    }
-
     const filesControllerAdapter = loadAdapter(filesAdapter, () => {
       return new GridStoreAdapter(databaseURI);
     });
     // Pass the push options too as it works with the default
     const pushControllerAdapter = loadAdapter(push && push.adapter, ParsePushAdapter, push || {});
-    const loggerControllerAdapter = loadAdapter(loggerAdapter, WinstonLoggerAdapter);
+
```
Comment (acinader) [inline src/ParseServer.js]: remove

**Comment by acinader** [comment]
I'll test it against https://github.com/acinader/parse-aws-firehose-logger-adapter/blob/master/README.md tomorrow morning.

File: src/Adapters/Logger/WinstonLogger.js (hunk 11 above)
Comment (flovilmart) [inline src/Adapters/Logger/WinstonLogger.js]: May fail if the folder exists, so we just let go

**Comment by facebook-github-bot-7** [comment]
@flovilmart updated the pull request.

File: src/ParseServer.js [hunk 13]
```diff
@@ -171,16 +168,13 @@ class ParseServer {
...
+    const loggerControllerAdapter = loadAdapter(loggerAdapter, WinstonLoggerAdapter, { jsonLogs, logsFolder, verbose, logLevel, silent });
```
Comment (steven-supersolid) [inline src/ParseServer.js]: Should we load earlier and assign logging at the same time so that any initialisation logs can use the custom logger and options too?

File: src/cli/cli-definitions.js [hunk 14]
```diff
@@ -193,6 +193,17 @@ export default {
     env: "JSON_LOGS",
     help: "Log as structured JSON objects"
   },
+  "logLevel": {
+    env: "PARSE_SERVER_LOG_LEVEL",
+    help: "Sets the level for logs"
+  },
+  "logsFolder": {
+    env: "PARSE_SERVER_LOGS_FOLDER",
+    help: "Folder for the logs (defaults to './logs')",
```
Comment (steven-supersolid) [inline src/cli/cli-definitions.js]: Perhaps add a note for using null to disable

File: src/ParseServer.js (hunk 13 above)
Comment (flovilmart) [inline src/ParseServer.js]: Not sure it would change much, but we could.

File: src/cli/cli-definitions.js (hunk 14 above)
Comment (flovilmart) [inline src/cli/cli-definitions.js]: Yup, and something to transform 'null' to null

**Comment by facebook-github-bot-9** [comment]
@flovilmart updated the pull request.

File: src/Adapters/Logger/WinstonLoggerAdapter.js (hunk 8 above)
Comment (flovilmart) [inline src/Adapters/Logger/WinstonLoggerAdapter.js]: I'd keep it separated to keep it clean

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by acinader** [comment]
my testing is not going well :).

I removed all logger configuration from my startup of parse-server.

Logs are getting written to ./logs as expected.

I am not able to use the logger to log.

in my cloud code I am trying to include logger like so:
```
const logger = require('parse-server').logger
```
but it isn't working (undefined).

How am I supposed to access the logger for writing?

**Comment by flovilmart** [comment]
In cloud code the logger is exposed through req.log. You should use that one. 

Also cloud code modules are loaded first, we could try to load them last.

**Comment by acinader** [comment]
So two other issues I'm not sure how to work around:

1.  I can no longer use the logger in my index:

```
const conf = require('config');
const express = require('express');
const ParseServer = require('parse-server').ParseServer;
const logger = require('parse-server').logger;
const playlist = require('./routes/playlist');
const sns = require('./routes/sns');
const app = express();
const api = new ParseServer(conf.parseServerOptions);

const port = conf.parseServerOptions.port;

app.use('/parse', api);

app.get('/ping', (req, res) => {
  res.send('pong');
});

app.get('/playlist/:hlsVersion/:storyId', playlist.getPlaylist);

app.post('/receiveSnsMessage', sns.receive);

app.listen(port, () => {
  console.log(`epic parse-server running on port ${port}.`);
});

module.exports = app;
```

2. I am unable to configure the logger in my conf files.  I'm using https://www.npmjs.com/package/config for config and my logger config looks like this:

```
....
loggerAdapter: defer(cfg => new FirehoseLoggerAdapter(cfg.firehoseAdapterOptions)),
....
firehoseAdapterOptions: {
    streamName: 'epic-logs',
    level: 'info',
  },
...
```
but it fails when trying to require WinstonLogger cause the defaults aren't set yet.

**Comment by flovilmart** [comment]
Try do the require `const logger = ParseServer.logger;` AFTER parse server init, before, it's not materialized as not instantiated :)

For your adapter: try:
```
{
  loggerAdapter: {
    module: 'firehose-logger-adapter-module'
    options: {} 
  }
}
``` 

The adapter loader should be able to handle it, and load your module correctly. 
I'm not sure what would cause the defaults not to be set (circular references maybe?) Can you provide the full trace?

**Comment by acinader** [comment]
wow, i'm just batting zero today ;)

in my index, i tried:

const api = new ParseServer(conf.parseServerOptions);
const logger = api.logger;

but logger is still undefined.

and on the adapterLoader, I tried:

```
    loggerAdapter: {
      module: 'parse-aws-firehose-logger-adapter',
      options: {
        streamName: 'epic-dev-logs',
        level: 'info',
      },
```

but it gives me:
```
Error: Cannot find module 'parse-aws-firehose-logger-adapter'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Module.require (module.js:468:17)
```

the module is listed as a dependency in my package.json 
```
✗ ls -d node_modules/parse-aws-firehose-logger-adapter 
node_modules/parse-aws-firehose-logger-adapter
```

**Comment by flovilmart** [comment]
```
const api = new ParseServer(conf.parseServerOptions);
const logger = api.logger;

```

The logger is not exposed here.

```
const ParseServer = require('parse-server').ParseServer;
const api = new ParseServer(conf.parseServerOptions);
const logger = require('parse-server').logger;
```

Does that do the trick?

For the other issue, I'm not sure why the adapter wouldn't load correctly :/ That's pretty odd...

**Comment by acinader** [comment]
yes, your incantation works in index.

also I can no include the logger in cloud code with 

```
const logger = require('parse-server').logger.
```

now works again in cloud code.

I'm assuming that my logging adapter isn't loading because of how I have defined/exported it, so working on that.

**Comment by facebook-github-bot-2** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@acinader just tested and it seems that we can load from a module from NPM. that's odd (or something that I don't take into account)

**Comment by acinader** [comment]
pretty sure that the reason it didn't work is cause i am npm linking parse-server.

when I configure it with a full path, it does load, but I get this error:

```
/Users/arthur/code/parse-server/lib/Controllers/AdaptableController.js:74
        throw new Error("Adapter prototype don't match expected prototype", adapter, mismatches);
        ^

Error: Adapter prototype don't match expected prototype
```

Interestingly, if i replace my adapter with:
```
      module: 'parse-server/lib/Adapters/Logger/WinstonLoggerAdapter',
```
I get the same error.

So, prior to this refactor, I was instantiating a logger adapter and passing it into the parse-server constructor.  I can no longer do that because defaults aren't set before I call the constructor for parse-server.  But now I am unable to pass in just the module name for either my class, or the winston adapter class. I modeled my adapter on WinstonLoggerAdapter, i.e. a class, export a default, and use new Class() to instantiate.

Looking into the adapter loader......

**Comment by acinader** [comment]
ok, i think i'm onto it.  i'll let you know if i have further questions....

**Comment by flovilmart** [comment]
> But now I am unable to pass in just the module name for either my class, or the winston adapter class. I modeled my adapter on WinstonLoggerAdapter, i.e. a class, export a default, and use new Class() to instantiate.

That should work correctly seems to be the loader that fails.
```
> /Users/arthur/code/parse-server/lib/Controllers/AdaptableController.js:74
        throw new Error("Adapter prototype don't match expected prototype", adapter, mismatches);
        ^

Error: Adapter prototype don't match expected prototype
```

Maybe because of the inheritance the `log` method is not seen on the adapter?

**Comment by acinader** [comment]
@flovilmart no, I'm trying it with WinstonLoggerAdapter not my adapter

```
loggerAdapter: {
      module: 'parse-server/lib/Adapters/Logger/WinstonLoggerAdapter',
      options: {
        level: 'info',
      },
    },
```

```
{ log: { expected: 'function', actual: 'undefined' } }
```

**Comment by flovilmart** [comment]
> You'd expect that I can configure the loggerAdapter like above, right?

 I just tested in a unit test, and it loads correctly.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@acinader does the last commit help?

File: src/defaults.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+  enableAnonymousUsers: true,
+  allowClientClassCreation: true,
+  maxUploadSize: '20mb',
+  verifyUserEmails: false,
+  preventLoginWithUnverifiedEmail: false,
+  sessionLength: 31536000,
+  expireInactiveSessions: true,
+  revokeSessionOnPasswordReset: true,
+  schemaCacheTTL: 5
+}
```
Comment (acinader) [inline src/defaults.js]: add cr

**Comment by acinader** [comment]
yes, pulling the defaults out enabled be to go back to how i was loading my adapter: instantiate the adapter and pass it in to parse-server constructor.

I'll try and poke around AdapterLoader.loadAdapter and AdaptableController.validateAdapter to figure out why WinstonLoggerAdapter wont load when configured via json like:

```
loggerAdapter: {
      module: 'parse-server/lib/Adapters/Logger/WinstonLoggerAdapter',
      options: {
        streamName: 'epic-dev-logs',
        level: 'info',
        jsonLogs: true,
      },
    },
```

**Comment by facebook-github-bot-6** [comment]
@flovilmart updated the pull request.

**Comment by flovilmart** [comment]
@acinader Yay! We're making progress :)

**Comment by flovilmart** [comment]
@acinader what if you pass the materialized adapter? does it work?

**Comment by acinader** [comment]
Yes, if I understand what materialized means....

This works:

```
const FirehoseLoggerAdapter = require('parse-aws-firehose-logger-adapter').FirehoseLoggerAdapter;
loggerAdapter: {
      module: FirehoseLoggerAdapter,
      options: { }
}
```

**Comment by flovilmart** [comment]
and that: `const FirehoseLoggerAdapter = require('parse-aws-firehose-logger-adapter').default;` ?

**Comment by acinader** [comment]
yeah that works too.

**Comment by flovilmart** [comment]
that's very odd indeed... I'll try in a few hours when back home

**Comment by flovilmart** [comment]
```
MODULE 56088: looking for "parse-aws-firehose-logger-adapter" in ["/Users/florent/src/Parse/parse-server/lib/Adapters/node_modules","/Users/florent/src/Parse/parse-server/lib/node_modules","/Users/florent/src/Parse/parse-server/node_modules","/Users/florent/src/Parse/node_modules","/Users/florent/src/node_modules","/Users/florent/node_modules","/Users/node_modules","/node_modules","/Users/florent/.node_modules","/Users/florent/.node_libraries","/usr/local/lib/node"]
```

That sucks ...

**Comment by flovilmart** [comment]
@acinader I managed to get it running:

in your project (alongside the topmost package.json)

```
git clone https://github.com/parseplatform/parse-server
cd parse-server
git checkout logging-refactor
cd ..
npm link ./parse-server
cd node-modules/parse-aws-firehose-logger-adapter
# link to the same parse-server
npm link ../../parse-server
cd ../../
# you should be good with that now :)
```

File: src/defaults.js (hunk 15 above)
Comment (flovilmart) [inline src/defaults.js]: done

**Comment by acinader** [comment]
ok.  learning more about npm linking.  thanks.  works for me too.

**Comment by flovilmart** [comment]
@acinader AWESOME!

**Comment by flovilmart** [comment]
@drew-gross can you give the final go/nits on that one?

File: src/defaults.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  silent: false,
+  enableAnonymousUsers: true,
+  allowClientClassCreation: true,
+  maxUploadSize: '20mb',
+  verifyUserEmails: false,
+  preventLoginWithUnverifiedEmail: false,
+  sessionLength: 31536000,
+  expireInactiveSessions: true,
+  revokeSessionOnPasswordReset: true,
+  schemaCacheTTL: 5
```
Comment (steven-supersolid) [inline src/defaults.js]: I think this value is incorrect (carried over from previous code), cache times are in milliseconds in InMemoryCache so should be 5000 if 5 seconds is the intention.
SchemaCache has a similar bug in the constructor but could be a separate PR

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@flovilmart updated the pull request.

**Comment by jpv123** [comment]
Any guide for begginers (like me) to setup this correctly? I'm a little bit tired of the big ammount of logging. It's nearly impossible to get the main info that I need from the logs in this moment

**Comment by acinader** [comment]
@jpv123 you can keep an eye on #2501 

I'm going to try and make the logging just like it was on parse.com.  I intend to get to it this week (unless someone beats me to it).

**Comment by sprabs** [comment]
Sorry, for the trouble, but I'm pretty confused at this point with logging on parse-server and things have changed a lot without any update to the wiki, so I'm not sure how to proceed. I thought this would be an appropriate place to comment since I noticed a back-and-forth about perhaps some similar confusion. 

Even with verbose set to true, I'm not seeing any logs (after the series of changes to parse-server logs changes). I'm running my parse server on AWS Elastic Beanstalk.

My index.js looks something like this:

```
var logger = require('parse-server/lib/Adapters/Logger/FileLoggerAdapter').FileLoggerAdapter;

// ...
var api = new ParseServer({
  databaseURI: databaseUri || 'mongodb://localhost:27017/dev',
  cloud: process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js',
  appId: process.env.APP_ID || 'myAppId',
  masterKey: process.env.MASTER_KEY || '', //Add your master key here. Keep it secret!
  serverURL: process.env.SERVER_URL || 'http://localhost:1337',  // Don't forget to change to https if needed
  loggerAdapter: new logger({ logsFolder: './' })
});
```

What do I need to fix to get logging back?

**Comment by flovilmart** [comment]
You don't have to pass loggerAdapter. 
Just set logsFolder and verbose upon initialization of the ParseServer object

**Comment by sprabs** [comment]
After following this thread, I tried this in my index.js:

```
var express = require('express');
var ParseServer = require('parse-server').ParseServer;

var databaseUri = process.env.DATABASE_URI || process.env.MONGOLAB_URI;

if (!databaseUri) {
  console.log('DATABASE_URI not specified, falling back to localhost.');
}

var WinstonLoggerAdapter = require('parse-server/lib/Adapters/Logger/WinstonLoggerAdapter');
var api = new ParseServer({
  databaseURI: databaseUri,
  appId: process.env.APP_ID || 'myAppId',
  masterKey: process.env.MASTER_KEY || '', //Add your master key here. Keep it secret!
  serverURL: process.env.SERVER_URL || 'http://localhost:1337',  // Don't forget to change to https if needed
  webhookKey: process.env.WEBHOOK_KEY || 'hook',
  push: {
    android: {
        senderId: process.env.GCM_SENDER_ID,
        apiKey: process.env.GCM_API_KEY
    },
    ios: {
        pfx: 'something.p12',
        bundleId: 'something.dev',
        production: false
    }
  },
  verbose: process.env.VERBOSE,
  logsFolder: process.env.PARSE_SERVER_LOGS_FOLDER,
  loggerAdapter: {
      module: WinstonLoggerAdapter,
      options: {
        level: 'info',
        jsonLogs: true,
      }
  },
  cloud: process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js'
});

var logger = require('parse-server').logger;

var app = express();

// Serve the Parse API on the /parse URL prefix
var mountPath = process.env.PARSE_MOUNT || '/1';
app.use(mountPath, api);

// Parse Server plays nicely with the rest of your web routes
app.get('/', function(req, res) {
  res.status(200).send('I dream of being a web site.');
});

var port = process.env.PORT || 1337;
app.listen(port, function() {
    console.log('parse-server running on port ' + port + '.');
});

```

And I added this at the top of my main.js:

```
var logger = require('../../src').logger;
console.log('logger:', logger); // OK
```

I am getting this error: 

```
> node index.js

/var/app/current/node_modules/parse-server/lib/Controllers/AdaptableController.js:74
        throw new Error("Adapter prototype don't match expected prototype", adapter, mismatches);
        ^

Error: Adapter prototype don't match expected prototype
    at LoggerController.validateAdapter (/var/app/current/node_modules/parse-server/lib/Controllers/AdaptableController.js:74:15)
    at LoggerController.set (/var/app/current/node_modules/parse-server/lib/Controllers/AdaptableController.js:80:12)
    at LoggerController.AdaptableController (/var/app/current/node_modules/parse-server/lib/Controllers/AdaptableController.js:39:18)
    at new LoggerController (/var/app/current/node_modules/parse-server/lib/Controllers/LoggerController.js:50:85)
    at new ParseServer (/var/app/current/node_modules/parse-server/lib/ParseServer.js:266:28)
    at new _ParseServer (/var/app/current/node_modules/parse-server/lib/index.js:36:16)
    at Object.<anonymous> (/var/app/current/index.js:16:11)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
```


What am I doing wrong? Also, for the path of logsFolder, is it an absolute or relative path?

**Comment by flovilmart** [comment]
Absolute or relative it doesn't matter. You can't specify both logsFolder/VERBOSE AND the loggerAdapter. Moreover, if you want verbose logs, the level should be 'verbose'

Setting verbose and logsFolder, will be passed to the default adapter which is the WinstonLoggerAdapter so you don't need to pass the loggerAdapter option. 

If you still want to pass the loggerAdapter option, remove logsFolder and verbose. Either do `new WinstonLoggerAdapter(options)` or

```
loggerAdapter: { module: pathToWinstonAdapter,
options: optionsForLogger }
```

I don't recommend the second way as we don't guarantee the path to the adapter will stay, nor the adapter will stay part of the core of parse-server the same way we removed the files adapters and push adapters.

I don't consider this a breaking change as those modules are internal and not supposed to be used outside of the package.

**Comment by flovilmart** [comment]
Setting the logsFolder option should set the dirname option of the Winston logger.

When running unit tests, we use that option to output the logs to the test_logs folder. When passing no options it should log to the ./logs folder. Does that work?

**Comment by sprabs** [comment]
Thanks @flovilmart! I now see the two folders: parse-server.err.2016-09-05, parse-server.info.2016-09-05

However, the error folder seems to be displaying even non-errors (with perfectly fine responses). Why do you think that's happening. Also, obviously, verbose = true is crazy. What log level do you recommend for debugging purposes and also in production and where do you set it?

**Comment by flovilmart** [comment]
There is a bug in the current version that makes both info and error log the same. This will be fixed in the next release

**Comment by flovilmart** [comment]
You should be able to set the level with the logLevel option directly on parse-server. Info level should be fine in production, you'll mostly have whatever was logged on good ol parse.com namely Cloud Code triggers and functions. 

VERBOSE is useful when you need to trace particular request/responses. Debug in only used with the Postgres Adapter to log all DB queries, unused with the mongo adapter.

**Comment by sprabs** [comment]
Ah, okay. I'm seeing the same issue on the dashboard as well (info and error being the same). Since I don't have the loggerAdapter line, how do I set a logLevel option of info in index.js?

Also, this is somewhat related: I did a pull of the latest parse server code and am installing it locally on my machine. When I do an npm install and deploy it, my parse server bugs out and crashes (complaining of an npm issue). I'm not sure what changed that would cause this.

**Comment by flovilmart** [comment]
Set logLevel directly in the options to create the ParseServer object. When in doubt look at the source code.

For the other thing, you should probably open another issue, as this as already turned into a discussion onto an issue that's already closed.

**Comment by sprabs** [comment]
Thanks, @flovilmart. 

logLevel didn't seem to work for me, but I'll try to check the documentation and see what I may be doing wrong. The iOS and Android clients still aren't getting error messages returned. I'm not sure if there's something else we need to set for that to make it the same as parse.com. When there are parse errors that are not javascript errors, the client doesn't get any kind of error back.

For the other problem I am having, I've filed an issue: https://github.com/ParsePlatform/parse-server/issues/2666

**Comment by flovilmart** [comment]
errors like object not found are not returned to the clients???

**Comment by sprabs** [comment]
No, instead it just has internal default errors from the SDK. For instance, for iOS, it can be something like: Error Domain=NSCocoaErrorDomain Code=3840 "No value." ...

**Comment by flovilmart** [comment]
that doesn't make any sense...

**Comment by sprabs** [comment]
Yeah, I agree. I'm really confused. Something seems to have broken in the last few weeks that is really making debugging anything impossible with parse-server. What do you suggest I check?

**Comment by flovilmart** [comment]
running with VERBOSE=1 and make sure whatever gets in / out is working correctly. I don't believe discussing on a PR that's been closed for a month is also the best way to convert issues. So please, open a proper issue, with the relevant informations to your issues.
